includes/class-edd-db-discounts.php

```diff
@@ -0,0 +1,460 @@
+
+					$end = date( 'Y-m-d H:i:s', strtotime( $args['start_date']['end'] ) );
+
+					$where .= " AND `start_date` <= '{$end}'";
+
+				}
+
+			} else {
+
+				$year  = date( 'Y', strtotime( $args['start_date'] ) );
+				$month = date( 'm', strtotime( $args['start_date'] ) );
+				$day   = date( 'd', strtotime( $args['start_date'] ) );
+
+				$where .= " AND $year = YEAR ( start_date ) AND $month = MONTH ( start_date ) AND $day = DAY ( start_date )";
+			}
+
+		}
+
+		if ( ! empty( $where ) ) {
+			$where = ' WHERE 1=1 ' . $where;
+		}
+
+		return $where;
+	}
+
+
+	/**
+	 * Count the total number of discounts in the database
+	 *
+	 * @access  public
+	 * @since   3.0
+	*/
+	public function count( $args = array() ) {
+		global $wpdb;
+
+		$where     = $this->parse_where( $args );
+		$cache_key = md5( 'edd_discounts_count' . serialize( $args ) );
+
+		$count = wp_cache_get( $cache_key, 'discounts' );
+
+		if( $count === false ) {
+
+			$sql   = "SELECT COUNT($this->primary_key) FROM " . $this->table_name . "{$where};";
+			$count = $wpdb->get_var( $sql );
+
+			wp_cache_set( $cache_key, $count, 'discounts', 3600 );
+
+		}
+
+		return absint( $count );
+	}
+
+	/**
+	 * Sets the last_changed cache key for customers.
+	 *
+	 * @access public
+	 * @since  3.0
+	 */
+	public function set_last_changed() {
+		wp_cache_set( 'last_changed', microtime(), $this->cache_group );
+	}
+
+	/**
+	 * Retrieves the value of the last_changed cache key for customers.
+	 *
+	 * @access public
+	 * @since  3.0
+	 */
+	public function get_last_changed() {
+		if ( function_exists( 'wp_cache_get_last_changed' ) ) {
+			return wp_cache_get_last_changed( $this->cache_group );
+		}
+
+		$last_changed = wp_cache_get( 'last_changed', $this->cache_group );
+		if ( ! $last_changed ) {
+			$last_changed = microtime();
+			wp_cache_set( 'last_changed', $last_changed, $this->cache_group );
+		}
+
+		return $last_changed;
+	}
+
+	/**
+	 * Create the table
+	 *
+	 * @access  public
+	 * @since   3.0
+	*/
+	public function create_table() {
+
+		global $wpdb;
+
+		require_once( ABSPATH . 'wp-admin/includes/upgrade.php' );
+
+		$sql = "CREATE TABLE " . $this->table_name . " (
+		id bigint(20) UNSIGNED NOT NULL AUTO_INCREMENT,
+		name varchar(200) NOT NULL,
+		code varchar(50) NOT NULL,
+		status varchar(20) NOT NULL,
+		type varchar(20) NOT NULL,
+		amount mediumtext NOT NULL,
+		description longtext NOT NULL,
+		max_uses bigint(20) NOT NULL,
+		use_count bigint(20) NOT NULL,
+		once_per_customer int(1) NOT NULL,
+		min_cart_price mediumtext NOT NULL,
+		product_reqs mediumtext NOT NULL,
+		product_condition varchar(3) NOT NULL DEFAULT 'all',
+		excluded_products mediumtext NOT NULL,
+		applies_globally int(1) NOT NULL DEFAULT 1,
+		created_date datetime NOT NULL DEFAULT '0000-00-00 00:00:00',
+		start_date datetime NOT NULL DEFAULT '0000-00-00 00:00:00',
+		end_date datetime NOT NULL DEFAULT '0000-00-00 00:00:00',
+		notes longtext NOT NULL,
+		PRIMARY KEY (id),
+		KEY code (code)
```

Comment:
we should have a `KEY` for the following ones as well:
`status (status)` -> will help with getting only discounts of a specific status
`end_date (end_date)` -> will help when doing our cleanup for expired discounts

Reply:
I have a lot of work to do on these still so please hold all feedback.